plugins/obs-outputs/rtmp-stream.c

```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+}
+
+static void adjust_bitrate(struct rtmp_stream *stream)
+{
+	obs_encoder_t *vencoder = obs_output_get_video_encoder(stream->output);
+	const char *encoder_id = obs_encoder_get_id(vencoder);
+	obs_data_t *params = obs_encoder_get_settings(vencoder);
+	int i_nal_hrd = strcmp(encoder_id, "obs_x264") == 0 ?
+			obs_data_get_int(params, "i_nal_hrd"): 0;
+
+	uint64_t cur_time_ms = os_gettime_ns() / 1000000;
+	uint64_t last_adjustment_time = stream->last_adjustment_time;
+
+	size_t counter = stream->congestion_counter;
+	stream->mean_congestion += (stream->congestion
+			- stream->congestion_array[counter]) / CONGESTION_ARRAY_SIZE;
+	stream->congestion_array[counter] = stream->congestion;
+	stream->congestion_counter = (stream->congestion_counter + 1) %
+			CONGESTION_ARRAY_SIZE;
+	float congestion = stream->congestion;
+	float last_congestion = stream->last_congestion;
+	float mean_congestion = stream->mean_congestion;
+	float max_congestion = find_maximum(stream->congestion_array,
+			CONGESTION_ARRAY_SIZE);
+
+	int current_bitrate = stream->dynamic_bitrate;
+	int initial_bitrate = stream->initial_bitrate;
+	int previous_bitrate;
+
+	float decrease_rate = (float)stream->bitrate_decrease_rate;
+	float increase_rate = (float)stream->bitrate_increase_rate;
+	float dynamic_threshold = 0.01 * (float)stream->dynamic_threshold;
+	uint64_t polling_time = 1000 * stream->recovery_polling_time; // in millisecs
+	uint64_t decrease_polling_time = stream->decrease_polling_time; // set in millisec in UI
+
+	/* X264_NAL_HRD_CBR=2 incompatible with dynamic variable bitrate.
+	 * On default settings, bitrate is adjusted downwards every second by 
+	 * about 15%.
+	 * Detection threshold of congestion is 0.15 on default but could be larger;
+	 * this would mean less reactivity but would allow micro-congestions
+	 * to heal themselves without changing too soon the bitrate.
+	 * We define a minimal bitrate at half the initial one.
+	 * The polling time tells how many milliseconds one waits before trying
+	 * to increase the bitrate.
+	 * The larger the polling time, the lower you want the threshold to be
+	 * in order to counter-balance the potential of many drops.
+	 * You also want the increase rate to be set at larger values.
+	 * Ex: if polling_time : 5 sec -> 30 sec
+	 * increase_rate : 10 % -> 30 - 40 %
+	 * threshold: 15 % congestion -> 5 %
+	 * decrease_polling_time: 1000 millisec -> 300 millisec
+	 */
+	bool decrease_br = congestion > dynamic_threshold &&
+			current_bitrate > (int)(initial_bitrate / 2) &&
+			(cur_time_ms - last_adjustment_time) > decrease_polling_time;
+
+	bool increase_br = mean_congestion < 0.05 && max_congestion < 0.15 &&
+			congestion < 0.05 && last_congestion < 0.05 &&
+			current_bitrate < initial_bitrate &&
+			(cur_time_ms - last_adjustment_time) > polling_time;
+
+	if (i_nal_hrd != 2 && stream->switch_variable_bitrate) {
+
+		if (decrease_br) {
+			current_bitrate = (int)(current_bitrate / (1.0 + decrease_rate/100.0));
```

Comment:
Style:  Math operators should always be surrounded by a single space on each side.